### ISSUE TITLE
Deprecate unpacking `dataclass` in favor of `namedtuple`

### ISSUE THREAD

**Issue by hendrikmakait** [dask/dask#9363]
As mentioned in https://github.com/dask/dask/pull/9345#issuecomment-1207863896, we currently unpack `dataclass`es to discover embedded Dask collections. However, the support for `dataclass`es is limited: 

>[W]e basically want to be able to unpack those `dataclass`es that do not have any custom `__init__`, `__post_init__`, feature fields with `init=False`, or use `InitVars`. Those `dataclass`es should ideally not be mutated, at the very least not here and mutation would likely create a host of problems down the line. 

In #9361, we add support for unpacking `namedtuple`, which seem to be a perfect match to the limited scope of functionality we want to support in unpacking `dataclass`es. They are immutable and have a trivial, unchangeable constructor that consists of all their fields with optional defaults.

Deprecating the unpacking of `dataclass`es in favor of `namedtuple` would give us a smooth way to remove traversal of `dataclass`es as mentioned in https://github.com/dask/dask/issues/6574#issuecomment-943411814 and help us get rid of a few issues: 
* Unclear boundary of support for unpacking `dataclass`es (e.g., unset `init=False` fields are fine, set `init=False` fields are not)
* Flawed approach to detecting un-supported `dataclass`es (i.e., detecting a custom `__init__` will likely remain imperfect)
* #5533
* Conceptually, this entire feature feels odd given the mutable nature of `dataclass`es

To deprecate unpacking of `dataclass`es, we can use the existing logic for unpacking to issue the `DeprecationWarning` whenever an embedded collection is detected.

**Comment by pavithraes** [comment]
Thank you, `@hendrikmakait`!

@jrbourbeau @ian-r-rose What do you think about this?

**Comment by ian-r-rose** [comment]
I agree that the issues around custom `__init__` functions and `init=False` fields make it difficult to fully support traversing, unpacking, and rebuilding dataclasses. I am a bit agnostic as to whether we need to deprecate it, or whether the fixes in #9345 are good enough moving forward.

Embedding collections within the field of a dataclass sounds like *extremely* niche behavior to me -- I'm not sure if anyone is using that today. One way we could find out is by adding a `DeprecationWarning` as you suggest, and seeing if anyone complains :) .

In any event, supporting `namedtuple`s well sounds like a good idea, and I think that #9361 could go in independently of whatever choice is made here.

**Comment by hendrikmakait** [comment]
> Embedding collections within the field of a dataclass sounds like extremely niche behavior to me -- I'm not sure if anyone is using that today. One way we could find out is by adding a `DeprecationWarning` as you suggest, and seeing if anyone complains :) .

@ian-r-rose: How do you feel about adding the `DeprecationWarning` now that `namedtuple` unpacking is supported? This might be a good chance to get rid of a rather inelegant piece of functionality or at least learn how those who complain about the deprecation are using that feature.
